WIP: separate onResponse handler, alternative to spawnAdapter, #23770

patriknw opened the pull request:
* Problems with spawnAdapter
  * lot's of boilerplate, mapping of messages, definition of additional internal messages
  * easy to create a resource leak by using spawnAdapter for each request, e.g. convenient
    to close over things and use them in the response mapping as is done in the IntroSpec,
    but that adapter is never stopped.
  * the mapping function must not access mutable state or ctx
* This makes it possible to have separate message handlers for response messages that doesn't
  conform to the actor's own message type. Doesn't have to be single response, can be like
  a subscription, or whatever interaction that is needed with other actors.
* I think that this together with the "safe" ask (see other PR in progress) could be a
  replacement for spawnAdapter. Wdyt?

Refs #23770

patriknw committed:
separate onResponse handler, alternative to spawnAdapter, #23770

* Problems with spawnAdapter
  * lot's of boilerplate, mapping of messages, definition of additional internal messages
  * easy to create a resource leak by using spawnAdapter for each request, e.g. convenient
    to close over things and use them in the response mapping as is done in the IntroSpec,
    but that adapter is never stopped.
  * the mapping function must not access mutable state or ctx
* This makes it possible to have separate message handlers for response messages that doesn't
  conform to the actor's own message type. Doesn't have to be single response, can be like
  a subscription, or whatever interaction that is needed with other actors.
* I think that this together with the "safe" ask (see other PR in progress) could be a
  replacement for spawnAdapter. Wdyt?

patriknw commented on akka-cluster-typed/src/test/scala/akka/cluster/ddata/typed/scaladsl/ReplicatorSpec.scala:
```diff
@@ -34,59 +34,47 @@ object ReplicatorSpec {
           msg match {
             case Increment ⇒
-              replicator ! Replicator.Update(Key, GCounter.empty, Replicator.WriteLocal, updateResponseAdapter)(_ + 1)
+              replicator ! Replicator.Update(Key, GCounter.empty, Replicator.WriteLocal, ctx.self.upcast)(_ + 1)
```
we might want something named more clear than the `upcast` for this purpose, e.g. `ctx.selfAs`

chbatey replied:
I find it weird it works given the type constraint on upcast, i imagine it wouldn't if you specified the type parameter explicitly

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/Actor.scala:
```diff
@@ -183,6 +183,14 @@ object Actor {
+        case _    ⇒ onMessage(ctx, msg.asInstanceOf[T])
+      }
+      new Immutable[T](x.asInstanceOf[(ActorContext[T], T) ⇒ Behavior[T]])
+    }
```
Is this really an improvement over `spawnAdapter`? We loose type safety and can still not tie in a per call context...

patriknw replied:
See the description of the PR. As far as I can see, `spawnAdapter` cannot be used for temporary adapters without high risk of resource leak, and they require boilerplate of defining internal messages.

For per call context we would have the "safe" ask.

johanandren replied:
I think I may have an idea to solve both now. I'll write down a note and try it out on wednesday.

johanandren commented:
After another round of discussion and thinking I think this is the best idea (that can actually work) so far.

chbatey commented on akka-cluster-typed/src/test/scala/akka/cluster/ddata/typed/scaladsl/ReplicatorSpec.scala:
```diff
@@ -34,59 +34,47 @@ object ReplicatorSpec {
+              Actor.unhandled // not dealing with failures
+          }
+        }
+        .onResponse[Replicator.Changed[GCounter]] { (ctx, chg) ⇒
```
i guess these will only be used for responses but there's now nothing stopping someone from casting an actor ref or using a type that is compatible with the actor type and using this for other mischievous things

patriknw replied:
that's true, lot's of nasty things can be done by a programmer ;-)
at least `ref.asInstanceOf[SomeOtherType]` smells pretty bad

chbatey reviewed (approved):
I created a few small examples and i do like the simplicity of this.

I do think we should add another method as upcast is misleading and the type constraint doesn't make sense for this purpose

chbatey commented:
I'd be interested to see what we could do with Dotty's union types for this in the far future

2m reviewed (approved):
LGTM. We could have a separate factory for the behaviour that has `onResponse` exposed. So the most usual `immutable` does not have this advanced feature exposed.

2m reviewed (approved):
LGTM. We could have a separate factory for the behaviour that has `onResponse` exposed. So the most usual `immutable` does not have this advanced feature exposed.

jrudolph commented on akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/Actor.scala:
```diff
@@ -183,6 +183,14 @@ object Actor {
     def onSignal(onSignal: PartialFunction[(ActorContext[T], Signal), Behavior[T]]): Behavior[T] =
       new BehaviorImpl.ImmutableBehavior(onMessage, onSignal)
+
+    def onResponse[M: ClassTag](handler: (ActorContext[T], M) ⇒ Behavior[T]): Immutable[T] = {
```
I guess it could be confusing that `ClassTag` does not distinguish between `Option[Int]` and `Option[String]`. If you have two response clauses for the same erased class, the first clause would always be taken and it would fail with a `ClassCastException` when an element with the second element type would arrive.

patriknw replied:
but general collection types should not be used as messages, not in typed nor in untyped

jrudolph replied:
Yes, that was just an example, it could also fail for all other type-parameterized values like the replication ones.

patriknw commented:
> We could have a separate factory for the behaviour that has onResponse exposed. So the most usual immutable does not have this advanced feature exposed.

@2m That's why I added it to `ImmutableBehavior`, which is the return type of `Behaviors.immutable`. Therefore it's optional, similar to `onSignal`. Did you have something else in mind?

jrudolph reviewed (commented):
I like the way of defining the additional `onResponse` handlers. I'm also a bit worried that the single mailbox gets now mingled for typed and untyped messages.

Could we create a single function ref for those internal messages that would wrap incoming messages with `InternalMessage` and only handle those in the `onResponse` / onInternalMessage` handler?

In the far future, that @2m talks about, we could even maintain another (union) type for internal messages so no casting would be necessary. Even until then it could make sense to allow giving a type for the internal messages even if that would be `AnyRef` for many cases. Hard to say how to build the API for that though...

2m commented:
> In the far future, that @2m talks about, we could even maintain another (union) type for internal messages so no casting would be necessary.

It was Christopher's far future. :D Anyway, in the near future wouldn't it be possible to gather the internal message types to an HList on every `onResponse` call? And then use that HList to guide the type inference of `ctx.selfAs` with something analogous to `Selector` from shapeless.

That would mean that to create such behavior all of the `onResponse` should be chained first, and then at the end `onMessage` would be called.

OFC no such nice API for Java.

> That's why I added it to ImmutableBehavior

I was thinking of having

```
final class WithAdditionalMessageTypes(...)
    extends BehaviorImpl.ImmutableBehavior(...) {

  def onResponse[M: ClassTag](handler: (ActorContext[T], M) ⇒ Behavior[T]): WithAdditionalMessageTypes[T] = {
        ...
    }

}
```

Which would also have a separate factory method in `Behaviors`. This would hide this advanced feature from the most common behavior - `Immutable`.

patriknw committed:
ctx.responseRef

* enforcing explicit response type to be able to detect if the
  protocol is changed, otherwise it would be inferred and the
  corresponding onResponse handler might not be defined

patriknw commented:
We have a few other ideas that I will try.

jrudolph commented:
Discussing it further highlighted a few aspects for me:

 * Interaction with other actors is an important part of actor programming. It seems the current proposal would partly give away type-safety for this part of actor programming, because the `selfAs` reply types are not tied to `onResponse` blocks, so if the target actor changes its protocol, it would not necessarily be noticed by the compiler. Another programmer error could be to use another type in `selfAs` as are actually handled with `onResponse` handlers.
 * The question is often whether and when to create real child actors for interaction with other actors. However, real child actors often add significant complexity because a child actor has its own mailbox and is executed concurrently with its parent. While that's sometimes intended it is also a source for race conditions. Often it would be nice to separate sessions of interacting with another actor, without having to spawn a real child, but being able to access state from the main actor. That could nicely be done with `spawnAdapter` but needs explicit life-cycle management of the spawned refs.
 * When you use just the immutable programming model and change states by switching to a new behavior, how do you clean up resources only needed by one state? Do you need to keep track of resources manually and release all resources not needed any more before switching to a new state?

patriknw commented:
After the discussions I came up with a better way: https://github.com/akka/akka/pull/24344
